v4/export/dump.go

```diff
@@ -196,193 +188,386 @@ func (d *Dumper) Dump() (err error) {
+				colTypes[i] = meta.ColumnTypes()[j]
+				break
+			}
 		}
 	}
-	// Do not dump table data and return nil
-	if conf.NoData {
-		return nil, nil
+	return pkFields, colTypes
+}
+
+func prepareTableListToDump(conf *Config, pool *sql.Conn) error {
+	databases, err := prepareDumpingDatabases(conf, pool)
+	if err != nil {
+		return err
 	}
 
-	if conf.Rows != UnspecifiedSize {
-		finished, chunksIterArray, err := concurrentDumpTable(ctx, conf, db, dbName, tableName)
-		if err != nil || finished {
-			return chunksIterArray, err
+	conf.Tables, err = listAllTables(pool, databases)
+	if err != nil {
+		return err
+	}
+
+	if !conf.NoViews {
+		views, err := listAllViews(pool, databases)
+		if err != nil {
+			return err
 		}
+		conf.Tables.Merge(views)
 	}
-	tableIR, err := SelectAllFromTable(conf, db, dbName, tableName)
+
+	filterTables(conf)
+	return nil
+}
+
+func dumpTableMeta(conf *Config, conn *sql.Conn, db string, table *TableInfo) (TableMeta, error) {
+	tbl := table.Name
+	selectField, _, err := buildSelectField(conn, db, tbl, conf.CompleteInsert)
 	if err != nil {
 		return nil, err
 	}
 
-	return []TableDataIR{tableIR}, nil
-}
+	var colTypes []*sql.ColumnType
+	// If all columns are generated
+	if selectField == "" {
+		colTypes, err = GetColumnTypes(conn, "*", db, tbl)
+	} else {
+		colTypes, err = GetColumnTypes(conn, selectField, db, tbl)
+	}
 
-func concurrentDumpTable(ctx context.Context, conf *Config, db *sql.Conn, dbName string, tableName string) (bool, []TableDataIR, error) {
-	// try dump table concurrently by split table to chunks
-	chunksIterCh := make(chan TableDataIR, defaultDumpThreads)
-	errCh := make(chan error, defaultDumpThreads)
-	linear := make(chan struct{})
-
-	ctx1, cancel1 := context.WithCancel(ctx)
-	defer cancel1()
-	var g errgroup.Group
-	chunksIterArray := make([]TableDataIR, 0)
-	g.Go(func() error {
-		splitTableDataIntoChunks(ctx1, chunksIterCh, errCh, linear, dbName, tableName, db, conf)
-		return nil
-	})
+	meta := &tableMeta{
+		database:      db,
+		table:         tbl,
+		colTypes:      colTypes,
+		selectedField: selectField,
+		specCmts: []string{
+			"/*!40101 SET NAMES binary*/;",
+		},
+	}
 
-Loop:
-	for {
-		select {
-		case <-ctx.Done():
-			return true, chunksIterArray, nil
-		case <-linear:
-			return false, chunksIterArray, nil
-		case chunksIter, ok := <-chunksIterCh:
-			if !ok {
-				break Loop
-			}
-			chunksIterArray = append(chunksIterArray, chunksIter)
-		case err := <-errCh:
-			return false, chunksIterArray, err
+	if conf.NoSchemas {
+		return meta, nil
+	}
+	if table.Type == TableTypeView {
+		viewName := table.Name
+		createTableSQL, createViewSQL, err := ShowCreateView(conn, db, viewName)
+		if err != nil {
+			return meta, err
 		}
+		meta.showCreateTable = createTableSQL
+		meta.showCreateView = createViewSQL
+		return meta, nil
 	}
-	if err := g.Wait(); err != nil {
-		return true, chunksIterArray, err
+	createTableSQL, err := ShowCreateTable(conn, db, tbl)
```

Comment:
This `err` is not caught.

Reply:
=_= Perhaps we need to set a linter for dumpling.

Reply:
Yes... Maybe we can file another PR to do this.